### PR TITLE
Improve AnyP::Uri::port_ and related port storage types

### PR DESCRIPTION
**Pull request by rousskov**
The old "unsigned short" type is used for various things all over the
code and cannot distinguish a parsed port zero from an absent port. In
theory, it could also hold values exceeding 65535, getting out of sync
with various (poorly duplicated) maximum port value checks.

No significant runtime changes are expected, but some transactions with
unknown request URI ports may now be reported correctly, showing "-"
instead of "0" for the missing port number. In some very special/rare
on_unsupported_protocol cases, Squid might no longer attempt to tunnel
non-HTTP requests to a zero port. Such attempts were failing anyway.

This change leaves many "unsigned short" port use cases untouched. Most
of them should be converted, but most (if not all) of those conversions
should be done in dedicated projects. Here are a few TODO highlights:

* ACLIntRange: Stores integers, is usable for integers, but assumes it
  is being configured with port numbers. We probably should enhance that
  class to handle "all" integers instead of expanding/deepening that
  rather limiting "input is just port numbers" assumption.

* CachePeer and internalRemoteUri(): Likely runtime changes related to
  squid.conf validity.

* Ftp::Server::listenForDataConnection() and friends: Triggers a few
  changes with runtime effects related to comm_local_port() failures.

* Ip::Address, Snmp::Session, fde: Too many low-level changes, some of
  which would be difficult to test.

* Config.Port.icp: Possible runtime changes related to squid.conf and
  command-line arguments validity.

Also used "%hu" instead of "%u" or "%d" to printf() ports. These changes
arguably improve code and might make some pedantic compiler happier, but
they do not fix any "real" bugs because variadic printf() arguments are
automatically promoted from short to int and, hence, printf()
implementations never get and never expect shorts.

### CODE REVIEW

File: src/anyp/Uri.cc
Comment (rousskov) [inline src/anyp/Uri.cc:369]: This diff may look as if this PR adds a "hack", but this PR does not change official code behavior. It only documents it. There is no bug here unless you consider a misleading `400/ERR_INVALID_URL` outcome a bug, and that misleading outcome problem existed before these changes.

File: src/acl/UrlPort.cc
Comment (rousskov) [inline src/acl/UrlPort.cc:17]: I have adjusted cf.data.pre to explicitly document that port zero in "port" ACL matches unknown ports. This is only a documentation improvement. The official ACL code (implicitly) works the same way.

File: src/anyp/Uri.cc
Comment (rousskov) [inline src/anyp/Uri.cc:579]: Normally, we get here with a port number. If there are some special cases when we get here without the port number, the old code made `authorityWithPort_` look like `host:0`. This PR leaves authorityWithPort_ unchanged -- `host`. I am not 100% sure whether that is an improvement, but it may be -- we are no longer leaking a bogus port number for somebody to pick up. If you disagree, I will adjust this code to append zero port again.

A _proper_ fix would be to leave `authorityWithPort_` completely empty and throw if somebody actually attempts to _use_ it later: Callers should not ask for and should not be given "authority with port" when there is no port. However, that proper  fix is outside this PR scope and it requires serious auditing work (and possibly significant code adjustments).

File: src/client_side.cc
Comment (rousskov) [inline src/client_side.cc:3187]: This minor change prevents possible bugs (in the official code) where we may try to tunnel to a zero port.

File: src/client_side_request.cc
Comment (rousskov) [inline src/client_side_request.cc:622]: The new `if` statement above (line 618) is needed so that we can safely dereference url.port() in the other `if` statement on line 621 and 623. AFAICT, this change preserves the original logic.

The official code here is rather messy/problematic and should be refactored, but this PR is not the right place to do that. If you want debugs() or hostHeaderVerifyFailed() parameters to say something else, please provide exact strings (while keeping the peculiarities of the surrounding official code in mind). Thank you.

File: src/tests/testHttpRequest.cc
Comment (rousskov) [inline src/tests/testHttpRequest.cc:50]: We cannot (easily) use AnyP::Port here because CPPUNIT_ASSERT_EQUAL()s below would not be able to "print" that. However, since the hard-coded port is indeed known in this case, this `KnownPort` usage is not wrong.

**Review by rousskov — COMMENTED** [review]
This PR is needed because #1253 is [blocked](https://github.com/squid-cache/squid/pull/1253#discussion_r1093732337) on these AnyP::Uri port improvements.

File: src/anyp/Uri.cc
Comment (yadij) [reply src/anyp/Uri.cc:369]: Side topic: do you have plans to do that conversion? or shall I resurrect my old project?

File: src/tests/testHttpRequest.cc
Comment (yadij) [reply src/tests/testHttpRequest.cc:50]: Please check that situation after adding an `AnyP::Port` stream print operator.

FWIW, I am okay with this as-is.

File: src/peer_select.cc
Comment (yadij) [inline src/peer_select.cc:533]: This `.value_or(0)` is broken. The remote port should never be _unknown_ at this point (though it may be `0` for localhost traffic).
If it is unset/unknown something has gone badly wrong with forwarding and the peer/server cannot be selected.

Please either fix (throw an error ?) , or add XXX to document the bug.

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/anyp/Uri.cc
Comment (rousskov) [reply src/anyp/Uri.cc:369]: I think the answer depends, in part, on what code you already have. Can you share a pointer to the corresponding git branch in your repository? To avoid misunderstanding, I am _not_ going to review that code -- I understand that it may be very stale and unpolished/unfinished -- I am only going to silently assess how far ahead it is and whether I agree with its overall direction, in order to answer your question.

File: src/acl/UrlPort.cc
Comment (rousskov) [reply src/acl/UrlPort.cc:17]: Not done. I disagree with some aspects of the suggested out-of-scope TODO and believe that planning this enhancement requires more discussion/research than we should do in this PR. Let's hash it out outside this PR if/when one of us starts working on `port` ACL enhancement.

File: src/format/Format.cc
Comment (rousskov) [reply src/format/Format.cc:511]: I am skipping this secondary change request until its [parent](https://github.com/squid-cache/squid/pull/1255/files#r1095565843) change request is resolved. FWIW, you can resolve this change request now because the _compiler_ will force me to make this adjustment after/if the code is adjusted the way that parent change request suggests.

File: src/tests/testHttpRequest.cc
Comment (rousskov) [reply src/tests/testHttpRequest.cc:50]: I am skipping this secondary change request until its [parent](https://github.com/squid-cache/squid/pull/1255/files#r1095576128) change request is resolved.

File: src/anyp/Uri.cc
Comment (rousskov) [reply src/anyp/Uri.cc:579]: Done at 5d9fe29.

File: src/anyp/Uri.h
Comment (rousskov) [reply src/anyp/Uri.h]: Done in commit f53d10b.

File: src/cf.data.pre
Comment (rousskov) [reply src/cf.data.pre]: Done in commit 3962b01.

File: src/client_side.h
Comment (rousskov) [reply src/client_side.h:486]: Done in commit c18ccde.

**Review by rousskov — COMMENTED** [review]
Thank you for a prompt review. I believe I have addressed all top-level change requests, satisfying most of them. Please re-review.

File: src/anyp/Uri.cc
Comment (eduard-bagdasaryan) [reply src/anyp/Uri.cc:579]: We can simplify a little:
```suggestion
            authorityWithPort_.appendf(":%hu", *port());
```
since we already checked with `has_value()`.

File: src/client_side.cc
Comment (eduard-bagdasaryan) [inline src/client_side.cc:3200]:
```suggestion
    auto http = buildFakeRequest(connectHost, *connectPort, payload);
```

File: src/anyp/Uri.h
Comment (eduard-bagdasaryan) [inline src/anyp/Uri.h:57]:
```suggestion
        port_.reset();
```

Comment (rousskov) [reply src/anyp/Uri.h:57]: Why do you think reset() is better here?

PR code feels slightly clearer to me -- it explicitly sets `port_` to a well-known "no value" state. The proposed change will result in the same "no value" state, of course, but it is slightly less clear IMO because the reader may not remember the `port_` type and might suspect that its reset() method changes the port to some known default value like 80 or 0.

Comment (eduard-bagdasaryan) [reply src/anyp/Uri.h:57]: There is no big difference, as you noted, but it seems to me that reset() is more in line with the calling clear() method - i.e., resetting all values to the 'default' state (equivalent to the default-constructed object). 

> a well-known "no value" state

Methods such as 'reset()' are common for STL types (e.g., smart pointers also have them). On the other hand,  'default' types (such as nullopt) are specific for each type.

Comment (rousskov) [reply src/anyp/Uri.h:57]: > Methods such as 'reset()' are common for STL types... On the other hand, 'default' types (such as nullopt) are specific for each type.

I agree that using the same (common) method for many data members is an advantage. IMO, being clear that there is no default port here outweighs that advantage in this std::optional context. We seem agree that the difference is small and unimportant, so if others prefer reset() here, I will accept your suggestion instead.

File: src/client_side.cc
Comment (rousskov) [reply src/client_side.cc]: Done in commit b6e56d4.

File: src/anyp/Uri.cc
Comment (yadij) [reply src/anyp/Uri.cc:369]: As far a I got was PR #335 to prepare the validation functions for using SBuf produced by a Tokenizer.

File: src/peer_select.cc
Comment (yadij) [reply src/peer_select.cc:533]: These Connection are intended to be (if not actually) opened. That requires a valid port number before it is sent to `handlePath()` below. The code leading up to peer selection is responsible for ensuring that the request-target is at least valid as a destination. Otherwise it is not worth even doing the IP lookup on the non-peer destinations.

(IP cache is a "red herring", this is the caller and obviously has access to the `request->url` destination details).

**Review by yadij — CHANGES_REQUESTED** [review]
AFAICS, just the `PeerSelector::noteIp(const Ip::Address &ip)` situation to resolve now.

File: src/peer_select.cc
Comment (rousskov) [reply src/peer_select.cc:536]: > Otherwise it is not worth even doing the IP lookup on the non-peer destinations.

In short, I believe there are two reasonable solutions here:

1. Change PeerSelector so that it never selects DIRECT destinations for request-targets with unknown ports, under the assumption that Squid does not have any use cases and does not support any protocols that would be able to use a portless DIRECT destination. This forwarding route selection logic change is clearly outside this PR scope, even if it does not result in any functionality changes for existing Squid installations.

2. Continue to let the caller decide what DIRECT selections are valid (for caller purposes). This is what the current PR code is already doing.

> > Please either fix (throw an error ?) , or add XXX to document the bug.

Added an XXX in commit 41334f0. Please unblock this PR.

File: src/anyp/Uri.cc
Comment (yadij) [reply src/anyp/Uri.cc:369]: Sure. All good.

File: src/client_side_request.cc
Comment (yadij) [reply src/client_side_request.cc:622]: I've spent a while looking this over now and does seem to be both correct and the cleanest current change.

FTR, status quo is what it is because of the incremental discovery of the "safe" or "only partially unsafe" cases in the sparse matrix of input permutations. A refactor risks obscuring which (and how) cases are handled, do not attempt expecting an easy or quick improvement.

File: src/peer_select.cc
Comment (yadij) [reply src/peer_select.cc:536]: Dislike, but will stop objecting.

Comment (yadij) [reply src/peer_select.cc:536]: Oh, as for out of scope: this is the non-tunnel equivalent piece of the forwarding bug half-fixed in this PR with reasoning https://github.com/squid-cache/squid/pull/1255/files/faeb96979ae6cfb0829a2676cb9751efeed48996#r1095191336

**Review by yadij — APPROVED** [review]
Thank you for addressing my concerns. Just the discussion with Eduard to go before merge.

**Comment by rousskov** [comment]
> Just the discussion with Eduard to go before merge.

@eduard-bagdasaryan, I am not aware of any ongoing or pending discussions on this PR, but if you are waiting for me, please let me know -- I could have missed it.

File: src/peer_select.cc
Comment (rousskov) [reply src/peer_select.cc:536]: > this is the non-tunnel equivalent

The critical difference between the two areas is that the high-level ConnStateData::initiateTunneledRequest() code already had logic for handling missing information -- that whole method is essentially about those missing info checks, and there was no question there that a port is required -- we know that our tunneling code needs ports. It was both easier and appropriate to rely on that existing logic to handle missing ports in that PR code.

In this low-level code, there is no existing framework to properly handle missing ports, no consensus that missing ports are actually a _problem_, and, after establishing that consensus, the missing code would need to be added in very different place(s) -- the problem cannot be properly addressed in _this_ method. That non-trivial out-of-scope change deserves a dedicated PR.

**Comment by eduard-bagdasaryan** [comment]
> I could have missed it.

It looks like all my concerns have been resolved.

**Pull request closed without merging**
